### DISCUSSION TITLE
Why only render a TextBlock when CreateContainer?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#17288] [Q&A] [answered]
I plan to make a VirtualizingWrapPanel.

First of all, I want to start with the simplest step and add an element to the UI by referring to the source code of VirtualizingStackPanel.

Here is my code:
```
 public class TestPanel : VirtualizingPanel, IScrollSnapPointsInfo
 {
     public bool AreHorizontalSnapPointsRegular { get; set; }
     public bool AreVerticalSnapPointsRegular { get; set; }

     List<Control> _VirtualizingElement = new List<Control>();

     public override void ApplyTemplate()
     {
         base.ApplyTemplate();
        
         this.Loaded += (sender, e) =>
         {
             var _item = Items[0];
             var generator = ItemContainerGenerator!;
             var container = generator.CreateContainer(Items[0], 0, "recycleKey");
             generator.PrepareItemContainer(container, "recycleKey", 0);
             AddInternalChild(container);
             generator.ItemContainerPrepared(container, "recycleKey", 0);                
         };
     }
     

     public event EventHandler<RoutedEventArgs> HorizontalSnapPointsChanged;
     public event EventHandler<RoutedEventArgs> VerticalSnapPointsChanged;



     protected override Control? ScrollIntoView(int index)
     {
         throw new NotImplementedException();
     }

     protected override Control? ContainerFromIndex(int index)
     {
         throw new NotImplementedException();
     }

     protected override int IndexFromContainer(Control container)
     {
         throw new NotImplementedException();
     }

     protected override IEnumerable<Control>? GetRealizedContainers()
     {
         throw new NotImplementedException();
     }

     protected override IInputElement? GetControl(NavigationDirection direction, IInputElement? from, bool wrap)
     {
         throw new NotImplementedException();
     }

     public IReadOnlyList<double> GetIrregularSnapPoints(Orientation orientation, SnapPointsAlignment snapPointsAlignment)
     {
         return new List<double>();
     }

     public double GetRegularSnapPoints(Orientation orientation, SnapPointsAlignment snapPointsAlignment, out double offset)
     {
         throw new NotImplementedException();
     }
 }
```

This is my test code
```
<ScrollViewer Name="MyScrollViewer">
	<ItemsControl Name="MyItemsControl" ItemsSource="{Binding TestOC}">
		<ItemsControl.ItemsPanel>
			<ItemsPanelTemplate>
				<Local:TestPanel></Local:TestPanel>
			</ItemsPanelTemplate>
		</ItemsControl.ItemsPanel>
		<ItemsControl.ItemTemplate>
			<DataTemplate x:DataType="Local:TestM">
				<Border Width="100" Background="{Binding Background}" Height="100">
					<TextBlock TextAlignment="Center" TextWrapping="Wrap" Width="50" Text="{Binding Content}" Foreground="White" VerticalAlignment="Center" HorizontalAlignment="Center"></TextBlock>
				</Border>
			</DataTemplate>
		</ItemsControl.ItemTemplate>
	</ItemsControl>
</ScrollViewer>
```

![image](https://github.com/user-attachments/assets/1ab24595-0520-4eb5-b65a-416cc9b2a041)

![image](https://github.com/user-attachments/assets/f12b845b-d0dd-491a-b417-91ddb4ce2332)

It's strange that after running, only the TextBlock is rendered in the Visual Tree, but the Border set in the DataTemplate is not rendered, and the content of the TextBlock is also wrong. What's wrong with my code?

**Comment by IanRawley** [comment, marked as answer]
I'm not an expert, but here: 
```
generator.PrepareItemContainer(container, "recycleKey", 0);
```
It looks like you're passing the string "recycleKey" as the object to place in the container.  Thus the content becomes the string "recycleKey", which doesn't match your data template and hits the default of just a TextBlock containing object.ToString().

Swap "recycleKey" in the PrepareItemContainer and ItemContainerPrepared calls with Items[0] and see what happens.
See [documentation here](https://github.com/AvaloniaUI/Avalonia/blob/58de42b101ae19bfd84891423d949ed7f346f777/src/Avalonia.Controls/Generators/ItemContainerGenerator.cs#L110), specifically: 
```
        /// <param name="item">The item to display.</param>
```

**Reply by kyuranger** [reply]
> I'm not an expert, but here:
> 
> ```
> generator.PrepareItemContainer(container, "recycleKey", 0);
> ```
> 
> It looks like you're passing the string "recycleKey" as the object to place in the container. Thus the content becomes the string "recycleKey", which doesn't match your data template and hits the default of just a TextBlock containing object.ToString().
> 
> Swap "recycleKey" in the PrepareItemContainer and ItemContainerPrepared calls with Items[0] and see what happens. See [documentation here](https://github.com/AvaloniaUI/Avalonia/blob/58de42b101ae19bfd84891423d949ed7f346f777/src/Avalonia.Controls/Generators/ItemContainerGenerator.cs#L110), specifically:
> 
> ```
>         /// <param name="item">The item to display.</param>
> ```

That's it. It works now!
